File: music-universe/music-universe-ui/src/music-universe/sources/lastfm/types/lastfm-track.ts
import {BaseLastfmEntity} from "./lastfm-entity";
import {type ArtistRelatedRawEntity} from "@/music-universe/shared/types/entities.ts";
import type {Track} from "@/music-universe/shared/types/entities.ts";
import type {LastfmArtistDto} from "@/music-universe/sources/lastfm/api/lastfm-artists.ts";
import type {ApprovalStatusType} from "@/music-universe/sources/lastfm/constants/approvalStatus.ts";
import type {LastfmArtist} from "@/music-universe/sources/lastfm/types/lastfm-artist.ts";

/**
 * LastFM Track entity that extends BaseRawEntity and implements LastfmEntity
 */
export class LastfmTrack
    extends
        BaseLastfmEntity<"track">
    implements
        ArtistRelatedRawEntity<"track">
{
    url: string;
    mbid: string | null;
    playCount: number | null;
    listenersCount: number | null;
    artist?: LastfmArtistDto;

    constructor(
        id: number,
        name: string,
        url: string,
        mbid: string | null,
        approvalStatus: ApprovalStatusType,
        playCount: number | null,
        listenersCount: number | null,
        artist?: LastfmArtist,
        masterEntity?: Track
    ) {
        super(id, name, approvalStatus, masterEntity);
        this.url = url;
        this.mbid = mbid;
        this.playCount = playCount;
        this.listenersCount = listenersCount;
        this.artist = artist;
    }

    getEntityType(): "track" {
        return "track";
    }

    getExternalArtistId(): number | undefined {
        return this.artist?.id;
    }

    getMasterArtistId(): number | undefined {
        return this.getMasterEntity()?.id;
    }
}

File: music-universe/music-data-raw-lastfm/src/main/java/yurykorzun/art/universe/music/data/raw/lastfm/collectable/track/entity/LastfmTrack.java
package yurykorzun.art.universe.music.data.raw.lastfm.collectable.track.entity;

import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.SuperBuilder;
import yurykorzun.art.universe.music.data.raw.lastfm.collectable.artist.entity.LastfmArtist;
import yurykorzun.art.universe.music.data.raw.lastfm.collectable.common.entity.BaseLastfmEntity;
import yurykorzun.art.universe.music.data.raw.lastfm.collectable.common.entity.LastfmEntityType;

import java.util.Objects;

@Entity(name = "track")
@SuperBuilder
@NoArgsConstructor
@Getter @Setter
public class LastfmTrack extends BaseLastfmEntity {

    @Id
    @SequenceGenerator(
        name = "track_seq_gen",
        sequenceName = "track_seq",
        allocationSize = 50
    )
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "track_seq_gen")
    @Setter(value = AccessLevel.NONE)
    private long id;

    @Column(name = "mbid")
    private String mbid;

    @NonNull
    @Column(name = "url")
    private String url;

    @Column(name = "duration")
    private Integer duration;

    @Deprecated
    @Column(name = "is_streamable")
    private Boolean isStreamable;

    @Column(name = "listeners_count")
    private Integer listenersCount;

    @Column(name = "play_count")
    private Long playCount;
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "artist_id")
    private LastfmArtist artist;

    @Override
    public LastfmEntityType getType() {
        return LastfmEntityType.TRACK;
    }

    @Override
    public String getUniqueKey() {
        return getUrl();
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof LastfmTrack that)) return false;
        if (!super.equals(o)) return false;
        return Objects.equals(duration,     that.duration)
            && Objects.equals(isStreamable,   that.isStreamable)
            && Objects.equals(mbid,         that.mbid)
            && Objects.equals(url,          that.url);
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), duration, isStreamable, mbid, url);
    }
}
